lib/Doctrine/DBAL/Platforms/AbstractPlatform.php

```diff
@@ -3174,6 +3211,16 @@ public function supportsPrimaryConstraints()
         return true;
     }
 
+    /**
+     * Whether the platform supports named primary key constraints.
+     *
+     * @return bool
+     */
+    public function supportsNamedPrimaryConstraints()
```

Comment:
I didn't think through this idea enough but it may be a good moment to start introducing platform-level interfaces for features like this. Right now, the features are structured in a way which is rather unfriendly to static analysis:
```php
function supportsTheFeature(): bool;

/**
 * @throws NotImplemented
 */
abstract function doTheFeature()
{
    throw new NotImplemented();
}

// client
if ($platform->supportsTheFeature()) {
    // we are not handling the exception which
    $platform->doTheFeature();
}
```

There's no way for the client to know whether or not to call a certain method on a certain platform and whether or not the documented exception needs to be handled at runtime.

Instead, all the feature's methods could be grouped under an interface so that the client knows that a feature can be used on the platform only if it implements the interface.

Later, the implementation of the features could be even moved out of the gigantic platform classes and composed.

Reply:
Tell me if I should reopen https://github.com/doctrine/dbal/pull/2839

Reply:
will use that as reference...

Reply:
@greg0ire it's not that black and white. I think the interfaces should be introduced only for something that can be done from the DBAL perspective on one platform and cannot be done on others (e.g. creating sequences). If something is doable in different ways on all platforms (e.g. column comments), the common interface should be defined in the abstract platform.

Reply:
@morozov would the result be that each platform implements 10-20 interfaces? How would that be different from the supports pattern if you do:

```php
if ($platform instanceof AllowsNamedPrimaryKeysPlatform) {
    $platform->doTheThing();
}
```
It would fatal error when not supported, so its also an "exception" and the instanceof check is essentialy the same as a `supports*` call.

Reply:
Let's proceed with the `supports()` for now. Having `if`-statements is the biggest problem regardless of whether it's `supports()` or `instanceof`. The client should be able to express their intent (`$platform->doStuff()`), not introspect its capabilities and work around the differences on the client side.

Reply:
hope to have some time to continue with this